src/java.base/share/classes/java/security/PEMEncoder.java

```diff
@@ -0,0 +1,320 @@
+     */
+    static public PEMEncoder of() {
+        return PEM_ENCODER;
+    }
+
+    /**
+     * Construct a String-based encoding based off the id type.
+     * @return the string
+     */
+    private String pemEncoded(PEMRecord pem) {
+        StringBuffer sb = new StringBuffer(1024);
+        sb.append("-----BEGIN ").append(pem.id()).append("-----");
+        sb.append(System.lineSeparator());
+        if (b64Encoder == null) {
+            b64Encoder = Base64.getMimeEncoder(64,
+                System.lineSeparator().getBytes());
+        }
+        sb.append(b64Encoder.encodeToString(
+            pem.pem().getBytes(StandardCharsets.ISO_8859_1)));
+        sb.append(System.lineSeparator());
+        sb.append("-----END ").append(pem.id()).append("-----");
+        sb.append(System.lineSeparator());
+        return sb.toString();
+    }
+
+    /**
+     * Encoded a given {@code DEREncodable} and return the PEM encoding in a
+     * String
+     *
+     * @param so a cryptographic object to be PEM encoded that implements
+     *           DEREncodable.
+     * @return PEM encoding in a String
+     * @throws IllegalArgumentException when the passed object returns a null
+     * binary encoding. An exception is thrown when PEMEncoder is
+     * configured for encryption while encoding a DEREncodable that does
+     * not support encryption.
+     * @throws NullPointerException when object passed is null.
+     * @see #withEncryption(char[])
+     */
+    public String encodeToString(DEREncodable so) {
+        Objects.requireNonNull(so);
+        return switch (so) {
+            case PublicKey pu -> build(null, pu.getEncoded());
+            case PrivateKey pr -> build(pr.getEncoded(), null);
+            case KeyPair kp -> {
+                if (kp.getPublic() == null) {
+                    throw new IllegalArgumentException("KeyPair does not " +
+                        "contain PublicKey.");
+                }
+
+                if (kp.getPrivate() == null) {
+                    throw new IllegalArgumentException("KeyPair does not " +
+                        "contain PrivateKey.");
+                }
+                yield build(kp.getPrivate().getEncoded(),
+                    kp.getPublic().getEncoded());
+            }
+            case X509EncodedKeySpec x -> build(null, x.getEncoded());
+            case PKCS8EncodedKeySpec p -> build(p.getEncoded(), null);
+            case EncryptedPrivateKeyInfo epki -> {
+                if (keySpec != null) {
+                    throw new IllegalArgumentException("encrypt was " +
+                        "incorrectly used");
+                }
+                try {
+                    yield pemEncoded(new PEMRecord(
+                        PEMRecord.ENCRYPTED_PRIVATE_KEY, epki.getEncoded()));
+                } catch (IOException e) {
+                    throw new SecurityException(e);
+                }
+            }
+            case Certificate c -> {
```

Comment:
Shouldn't this be `X509Certificate` and `X509CRL` for line 191 since those are the permitted `DerEncoded` subtypes? I'm assuming it would be an error (compiler?) to try to pass in other subtypes.

Reply:
Yes the upcoming update fixes this